pkg/process/checks/net.go

```diff
@@ -172,13 +203,74 @@ func (c *ConnectionsCheck) Run(nextGroupID func() int32, _ *RunOptions) (RunResu
 
 	log.Debugf("collected connections in %s", time.Since(start))
 
+	if c.statsd != nil {
+		_ = c.statsd.Gauge("datadog.process.connections.count", float64(len(conns.Conns)), []string{}, 1)
+		_ = c.statsd.Histogram("datadog.process.connections.collection_time", time.Since(start).Seconds(), []string{}, 1)
+	}
+
 	c.npCollector.ScheduleConns(conns.Conns, conns.Dns)
 
 	groupID := nextGroupID()
 	messages := batchConnections(c.hostInfo, c.maxConnsPerMessage, groupID, conns.Conns, conns.Dns, c.networkID, conns.ConnTelemetryMap, conns.CompilationTelemetryByAsset, conns.KernelHeaderFetchResult, conns.CORETelemetryByAsset, conns.PrebuiltEBPFAssets, conns.Domains, conns.Routes, conns.Tags, conns.AgentConfiguration, c.serviceExtractor)
 	return StandardRunResult(messages), nil
 }
 
+// shouldRunCheck determines whether the connections check should run based on configuration and system capacity.
+// It handles both traditional (always run) and dynamic interval (capacity-aware) modes.
+// Returns true if the check should run, false if it should be skipped.
+func (c *ConnectionsCheck) shouldRunCheck(start time.Time) bool {
+	if !c.useDynamicInterval {
+		// Traditional behavior: always run the check
+		status.UpdateLastCollectTime(start)
+		log.Debugf("running connections check (dynamic interval disabled)")
+		return true
+	}
+
+	// Dynamic interval mode: use capacity-aware logic
+	timeSinceLastRun := start.Sub(c.lastFullRunTime)
+	// Add a small tolerance (250ms) to account for differences in the check interval and the guaranteed run interval
+	guaranteedIntervalWithTolerance := c.guaranteedRunInterval - (time.Millisecond * 250)
+	isTimeForGuaranteedRun := c.lastFullRunTime.IsZero() || timeSinceLastRun >= guaranteedIntervalWithTolerance
+
+	isNearCapacity := false
+	if c.sysprobeClient != nil {
+		var capacityErr error
+		isNearCapacity, capacityErr = c.checkCapacity()
+		if capacityErr != nil {
+			_ = log.Warnf("failed to check system-probe connection capacity: %v. Proceeding based on time interval.", capacityErr)
+			isNearCapacity = false
+			if c.statsd != nil {
+				_ = c.statsd.Count("datadog.process.connections.capacity_check_errors", 1, []string{}, 1)
+			}
+		}
+	} else {
+		log.Debug("system probe client not available, skipping capacity check.")
+	}
+
+	// Decide whether to run the full check
+	shouldRunFullCheck := isTimeForGuaranteedRun || isNearCapacity
```

Comment:
What's a full check vs just a check? This could be confusing; I would just call this `shouldRunCheck`.

Reply:
agreed, updated